c/src/ml-api-service-db.cc

```diff
@@ -0,0 +1,268 @@
+
+/**
+ * @brief Get the instance object of MLServiceLevelDB
+ * @return IMLServiceDB* instance of MLServiceLevelDB
+ */
+static IMLServiceDB *
+get_instance ()
+{
+  if (!instance) {
+    instance = new MLServiceLevelDB (ML_DATABASE_PATH);
+    instance->connectDB ();
+  }
+
+  return instance;
+}
+
+/**
+ * @brief Construct a new MLServiceLevelDB object
+ * @param path database path
+ */
+MLServiceLevelDB::MLServiceLevelDB (std::string path)
+:  path (path)
+{
+  db_obj = nullptr;
+}
+
+/**
+ * @brief Destroy the MLServiceLevelDB object
+ */
+MLServiceLevelDB::~MLServiceLevelDB ()
+{
+}
+
+/**
+ * @brief Connect the level DB and initialize the private variables.
+ */
+void
+MLServiceLevelDB::connectDB ()
+{
+  char *err = nullptr;
+  leveldb_options_t *db_options;
+
+  db_options = leveldb_options_create ();
+  leveldb_options_set_create_if_missing (db_options, 1);
+
+  db_obj = leveldb_open (db_options, path.c_str (), &err);
+  if (err != nullptr) {
+    _ml_loge ("Error! Failed to open Database : %s", err);
+    leveldb_free (err);
+    throw std::runtime_error ("Failed to connectDB()!");
+  }
+
+  db_roptions = leveldb_readoptions_create ();
+  db_woptions = leveldb_writeoptions_create ();
+  leveldb_writeoptions_set_sync (db_woptions, 1);
+
+  return;
+}
+
+/**
+ * @brief Set the pipeline description with the given name.
+ * @note If the name already exists, the pipeline description is overwritten.
+ * @param[in] name Unique name to retrieve the associated pipeline description.
+ * @param[in] pipeline_description The pipeline description to be stored.
+ */
+void
+MLServiceLevelDB::setPipelineDescription (std::string name,
+    std::string pipeline_description)
+{
+  char *err = NULL;
+
+  leveldb_put (db_obj, db_woptions, name.c_str (), name.size (),
+      pipeline_description.c_str (), pipeline_description.size (), &err);
+  if (err != nullptr) {
+    _ml_loge ("Failed to put %s : %s", name.c_str (), err);
+    leveldb_free (err);
+    throw std::runtime_error ("Failed to setPipelineDescription()!");
+  }
+
+  return;
+}
+
+/**
+ * @brief Get the pipeline description with the given name.
+ * @param[in] name The unique name to retrieve.
+ * @param[out] pipeline_description The pipeline corresponding with the given name.
+ */
+void
+MLServiceLevelDB::getPipelineDescription (std::string name,
+    std::string & pipeline_description)
+{
+  char *err = NULL;
+  char *value = NULL;
+  gsize read_len;
+
+  value = leveldb_get (db_obj, db_roptions, name.c_str (), name.size (),
+      &read_len, &err);
+  if (!value) {
+    _ml_loge ("Fail to find the key %s", name.c_str ());
+    leveldb_free (err);
+    throw std::invalid_argument ("Fail to find the key");
+  }
+
+  if (err != nullptr) {
+    _ml_loge ("Failed to get %s : %s", name.c_str (), err);
+    leveldb_free (err);
+    throw std::runtime_error ("Failed to getPipelineDescription()!");
+  }
+
+  pipeline_description = std::string (value, read_len);
+  return;
```

Comment:
Is there a reason to return? For the readability?

Reply:
In this case, just readability of code.